### PR TITLE
Fix the incorrect progress information when AdaptiveRepeatDataHook is enabled

### PR DESCRIPTION
**Pull request by sungmanc**
### Summary

This PR resolves an incorrect logging from the `OTXProgressHook`. Since `OTXSampler` + `AdaptiveRepeatDataHook` enlarges the total iterations, it should be applied before calling the `OTXProgressHook`. 

So, I added `get_proper_repeat_times` function and used this function `before_run` to change the runner's max_iter.

This PR tackles the incorrect log for the classification task and enable RepeatData to multi-label and H-label. For the other tasks, I'll make another PR.

Since the LR scheduler was changed, I attached the experimental results. Regression datasets for multi-label, h-label were used.
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>

<div style='direction:ltr'>

  |   | EfficientNet-b0 | Deit-tiny | MobileNet_v3_Large
-- | -- | -- | -- | --
Multi-label   (Before) | E2E   time | 0:00:54 | 0:00:49 | 0:00:40
  | Score | 96.51 | 95.67 | 96.70
H-label   (Before) | E2E   time | 0:01:04 | 0:00:40 | 0:00:39
  | Score | 72.72 | 73.62 | 73.03
Multi-label   (After) | E2E   time | 0:01:32 | 0:00:47 | 0:00:33
  | Score | 96.50 | 95.04 | 96.17
H-label   (After) | E2E   time | 0:01:04 | 0:00:41 | 0:00:33
  | Score | 72.37 | 73.62 | 72.64

</div>

</body>

</html>

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algorithms/common/adapters/mmcv/configurer.py [hunk 1]
```diff
@@ -413,6 +413,17 @@ def configure_hooks(
...
+        # Update adaptive repeat
+        if not self.training:
+            remove_from_configs_by_type(cfg.custom_hooks, "AdaptiveRepeatDataHook")
+        for custom_hook in cfg.custom_hooks:
+            if custom_hook["type"] == "AdaptiveRepeatDataHook":
+                data_cfg = cfg.get("data", {})
+                bs = data_cfg.get("train_dataloader", {}).get("samples_per_gpu", None)
+                bs = bs if bs is not None else data_cfg.get("samples_per_gpu", 0)
+                custom_hook["train_batch_size"] = bs
+                custom_hook["n_train_data"] = len(data_cfg.get("train", {}).get("otx_dataset", []))
```
Comment (goodsong81) [inline src/otx/algorithms/common/adapters/mmcv/configurer.py]: We have `self.get_data_cfg(cfg, "train")` for this.

File: src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py [hunk 2]
```diff
@@ -17,38 +18,57 @@ class AdaptiveRepeatDataHook(Hook):
...
     Args:
+        train_batch_size (int) : The batch size of the train dataloader
+        n_train_data (int) : The number of the training dataset
         coef (float, optional) : coefficient that effects to number of repeats
                        (coef * math.sqrt(num_iters-1)) +5
         min_repeat (float, optional) : minimum repeats
     """
 
-    def __init__(self, coef: float = -0.7, min_repeat: float = 1.0):
+    def __init__(self, train_batch_size: int, n_train_data: int, coef: float = -0.7, min_repeat: float = 1.0):
```
Comment (goodsong81) [inline src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py]: Matter of taste, but `train_batch_size` & `train_data_size` are not that bad combination.
Or `num_train_data` also is a good choice.

Personally, n_??? thing reminds me of old-fashioned C coding.

Just a nitpick, though :)

File: src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py [hunk 3]
```diff
@@ -79,6 +79,7 @@ def train_func_single_iter(batch_size):
         if idx_hooks_to_remove:
             idx_hooks_to_remove.sort()
             for i in reversed(idx_hooks_to_remove):
+                print(f"{copied_cfg.custom_hooks[i]} deleted.")
```
Comment (goodsong81) [inline src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py]: Maybe to be deleted?

File: src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py [hunk 4]
```diff
@@ -127,6 +127,10 @@ def _set_batch_size(cfg, batch_size: int):
     else:
         cfg.data.train_dataloader["samples_per_gpu"] = batch_size
 
+        for custom_hook in cfg.custom_hooks:
+            if custom_hook["type"] == "AdaptiveRepeatDataHook":
+                custom_hook["train_batch_size"] = batch_size
+
```
Comment (goodsong81) [inline src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py]: Why don't you use `update_or_add_custom_hook()`?

**Review by goodsong81 — COMMENTED** [review]
Generally, looks good. Could you check my comments?

File: src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py [hunk 5]
```diff
@@ -977,3 +973,25 @@ def area(x):
...
+    batch_size: int,
+    coef: float,
+    min_repeat: float,
+) -> float:
+    """Get proper repeat times for adaptive training.
+
+    Args:
+        n_data (int): The total number of the training dataset
+        batch_size (int): The batch size for the training data loader
+        coef (float, optional) : coefficient that effects to number of repeats
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py]:
```suggestion
        coef (float) : coefficient that effects to number of repeats
```
It seems typo.

File: src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py [hunk 6]
```diff
@@ -977,3 +973,25 @@ def area(x):
...
+                       (coef * math.sqrt(num_iters-1)) +5
+        min_repeat (float, optional) : minimum repeats
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py]:
```suggestion
        min_repeat (float) : minimum repeats
```
Same as above.

File: src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py [hunk 7]
```diff
@@ -17,38 +18,57 @@ class AdaptiveRepeatDataHook(Hook):
...
+        if self.n_repeats > 1:
+            iter_per_epoch = int(self.n_train_data / self.train_batch_size)
+
+            logger.info("Adaptive repeat is enabled")
+            logger.info(f"- Repeat times: {self.n_repeats}")
+            logger.info(f"- Batch size: {self.train_batch_size}")
+            logger.info(f"- Num iters per epoch: {iter_per_epoch} -> {iter_per_epoch * self.n_repeats}")
+            logger.info(f"- Total iters: {runner.max_iters} -> {runner.max_iters * self.n_repeats}")
+
+            runner._max_iters = int(runner.max_iters * self.n_repeats)
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py:53]: It uses protected attirbute in a class. Aren't there any ways to avoid it?

File: src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py [hunk 8]
```diff
@@ -17,38 +18,57 @@ class AdaptiveRepeatDataHook(Hook):
...
+
     def before_epoch(self, runner):
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py:55]: It's just a question. Should we make a new data loader every epoch? Can't we do it at first epoch once?

File: src/otx/algorithms/common/adapters/mmcv/configurer.py [hunk 9]
```diff
@@ -413,6 +413,17 @@ def configure_hooks(
         if hasattr(cfg, "algo_backend"):
             self._update_caching_modules(cfg)
 
+        # Update adaptive repeat
+        if not self.training:
+            remove_from_configs_by_type(cfg.custom_hooks, "AdaptiveRepeatDataHook")
+        for custom_hook in cfg.custom_hooks:
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/configurer.py]:
```suggestion
        if not self.training:
            remove_from_configs_by_type(cfg.custom_hooks, "AdaptiveRepeatDataHook")
            return
        for custom_hook in cfg.custom_hooks:
```
How about exit the functon if it isn't training?

File: src/otx/algorithms/common/adapters/mmcv/configurer.py [hunk 10]
```diff
@@ -413,6 +413,17 @@ def configure_hooks(
...
+            if custom_hook["type"] == "AdaptiveRepeatDataHook":
+                data_cfg = cfg.get("data", {})
+                bs = data_cfg.get("train_dataloader", {}).get("samples_per_gpu", None)
+                bs = bs if bs is not None else data_cfg.get("samples_per_gpu", 0)
+                custom_hook["train_batch_size"] = bs
+                custom_hook["n_train_data"] = len(data_cfg.get("train", {}).get("otx_dataset", []))
+
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/configurer.py]:
```suggestion
                custom_hook["n_train_data"] = len(data_cfg.get("train", {}).get("otx_dataset", []))
                break

```
How about to break to prevent it from iterating meaninglessly?

File: src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py [hunk 11]
```diff
@@ -977,3 +973,25 @@ def area(x):
         input_size = self.select_closest_size(input_size, input_size_preset)
         logger.info(f"-> Closest preset: {input_size}")
         return input_size
+
+
+def get_proper_repeat_times(
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py:977]: Could you add unit test for this function?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work! I left some comments. Please take a look.

File: src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py (hunk 8 above)
Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py:55]: Done. https://github.com/openvinotoolkit/training_extensions/pull/2510/commits/e1cf24fe04c36af2963f098052c248475a68e2b7

File: src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py (hunk 7 above)
Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py:53]: For now, yes. We should use that attribute.

File: src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py (hunk 4 above)
Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py]: Done. https://github.com/openvinotoolkit/training_extensions/pull/2510/commits/e1cf24fe04c36af2963f098052c248475a68e2b7

File: src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py (hunk 3 above)
Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py]: Done. https://github.com/openvinotoolkit/training_extensions/pull/2510/commits/e1cf24fe04c36af2963f098052c248475a68e2b7

File: src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py (hunk 11 above)
Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py:977]: I think this test includes this functionality. 
https://github.com/openvinotoolkit/training_extensions/blob/e1cf24fe04c36af2963f098052c248475a68e2b7/tests/unit/algorithms/common/adapters/mmcv/hooks/test_adaptive_repeat_data_hook.py#L33

File: src/otx/algorithms/common/adapters/mmcv/configurer.py [hunk 12]
```diff
@@ -416,13 +416,15 @@ def configure_hooks(
...
             remove_from_configs_by_type(cfg.custom_hooks, "AdaptiveRepeatDataHook")
+            return
         for custom_hook in cfg.custom_hooks:
             if custom_hook["type"] == "AdaptiveRepeatDataHook":
                 data_cfg = cfg.get("data", {})
                 bs = data_cfg.get("train_dataloader", {}).get("samples_per_gpu", None)
                 bs = bs if bs is not None else data_cfg.get("samples_per_gpu", 0)
                 custom_hook["train_batch_size"] = bs
-                custom_hook["n_train_data"] = len(data_cfg.get("train", {}).get("otx_dataset", []))
+                custom_hook["train_data_size"] = len(self.get_data_cfg(cfg, "train"))
```
Comment (goodsong81) [inline src/otx/algorithms/common/adapters/mmcv/configurer.py]: `len(train_data_cfg)` might be not what you want.

File: src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py (hunk 11 above)
Comment (eunwoosh) [reply src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py:977]: Yes it is in some degree. But it's insufficient for testing various situation, I think. For example, it can be better by having a code dealing with a wrong argument value in this function with those test cases.

File: src/otx/algorithms/common/adapters/mmcv/configurer.py (hunk 12 above)
Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/configurer.py]: Thanks :) I edited.

File: src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py (hunk 11 above)
Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py:977]: Added unit test to check the special cases (i.e. train_data_size = 0)

File: src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py (hunk 7 above)
Comment (eunwoosh) [reply src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py:53]: Ok than could you add some comments to explain why it's necessary?

File: src/otx/algorithms/common/adapters/mmcv/configurer.py (hunk 12 above)
Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/configurer.py]: For some reason, I decided to avoid using `get_data_cfg`. When `cfg` doesn't include the `data` field, `get_data_cfg` raises an error. However, there are some unit tests regarding the `custom_hook` that don't consider the `cfg.data`. So, it raises an error. To avoid it, I think using the data_cfg.get("train", {}).~~ is the better choice.

File: src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py (hunk 7 above)
Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/hooks/adaptive_repeat_data_hook.py:53]: I added it

File: src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py (hunk 3 above)
Comment (eunwoosh) [reply src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py]: I think he said this print statement can be deleted, is it right @goodsong81? And if you think this is necessary, I think `logger.debug` is better than `print`.

File: src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py [hunk 13]
```diff
@@ -126,6 +127,10 @@ def _set_batch_size(cfg, batch_size: int):
         cfg.data.videos_per_gpu = batch_size
     else:
         cfg.data.train_dataloader["samples_per_gpu"] = batch_size
+        update_or_add_custom_hook(cfg, {"type": "AdaptiveRepeatDataHook", "train_batch_size": batch_size})
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py]: Are there any reasons why `AdaptiveRepeatDataHook` is added? Isn't is sufficient to change batch size value if `AdaptiveRepeatDataHook` exists?

Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py]: I forgot the exact reason, however, if we didn't include the `AdaptieRepeatDataHook`, it raised an error at some cases.

File: src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py (hunk 3 above)
Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py]: removed it.

File: src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py (hunk 13 above)
Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py]: If auto_bs is enabled (i-seg, s-seg), the batch size is changed by automatic_bs. So, the batch size of the `AdaptiveRepeatDataHook` also should be changed. That's the reason why I added this line.

Comment (sungmanc) [reply src/otx/algorithms/common/adapters/mmcv/utils/automatic_bs.py]: Changed.

**Review by eunwoosh — DISMISSED** [review]

**Review by eunwoosh — APPROVED** [review]

**Review by goodsong81 — APPROVED** [review]
